src/csharp/Microsoft.Spark/Interop/Ipc/CallbackConnection.cs

```diff
@@ -0,0 +1,276 @@
+                    $"[{ConnectionId}] Finished running {_numCallbacksRun} callback(s).");
+            }
+        }
+
+        private void Stop()
+        {
+            _isRunning = false;
+            s_logger.LogInfo($"[{ConnectionId}] Stopping CallbackConnection.");
+        }
+
+        /// <summary>
+        /// Process the input and output streams.
+        /// </summary>
+        /// <param name="inputStream">The input stream.</param>
+        /// <param name="outputStream">The output stream.</param>
+        /// <param name="readComplete">True if stream is read completely, false otherwise.</param>
+        /// <returns>The connection status.</returns>
+        private ConnectionStatus ProcessStream(
+            Stream inputStream,
+            Stream outputStream,
+            out bool readComplete)
+        {
+            readComplete = false;
+
+            try
+            {
+                byte[] requestFlagBytes = SerDe.ReadBytes(inputStream, sizeof(int));
+                // For socket stream, read on the stream returns 0, which
+                // SerDe.ReadBytes() returns as null to denote the stream is closed.
+                if (requestFlagBytes == null)
+                {
+                    return ConnectionStatus.SOCKET_CLOSED;
+                }
+
+                // Check value of the initial request. Expected values are:
+                // - CallbackFlags.CLOSE
+                // - CallbackFlags.CALLBACK
+                int requestFlag = BinaryPrimitives.ReadInt32BigEndian(requestFlagBytes);
+                if (requestFlag == (int)CallbackFlags.CLOSE) {
+                    return ConnectionStatus.REQUEST_CLOSE;
+                }
+                else if (requestFlag != (int)CallbackFlags.CALLBACK)
+                {
+                    throw new Exception(
+                        string.Format(
+                            "Unexpected callback flag received. Expected: {0}, Received: {1}.",
+                            CallbackFlags.CALLBACK,
+                            requestFlag));
+                }
+
+                // Use callback id to get the registered handler.
+                int callbackId = SerDe.ReadInt32(inputStream);
+                if (!_callbackHandlers.TryGetValue(
+                        callbackId,
+                        out ICallbackHandler callbackHandler))
+                {
+                    throw new Exception($"Unregistered callback id: {callbackId}");
+                }
+
+                s_logger.LogInfo(
+                    string.Format(
+                        "[{0}] Received request for callback id: {1}, callback handler: {2}",
+                        ConnectionId,
+                        callbackId,
+                        callbackHandler));
+
+                // Run callback handler.
+                callbackHandler.Run(inputStream);
+
+                // Check the end of stream.
+                int endOfStream = SerDe.ReadInt32(inputStream);
```

Comment:
should we validate this before running callback?

Reply:
Calling Run() after `END_OF_STREAM` is read.